File: libdriveless/tests/python/test_angle.py
import sys, time
sys.path.append("../../")
sys.path.append("../")
import unittest, math
from pydriveless import angle
from test_utils import AssertTolerance

class TestAngle(unittest.TestCase):

    def test_boolean_operators(self):
        a = angle.new_deg(12.2)
        v = angle.new_deg(12.2)
        
        self.assertTrue(a == v)
        self.assertTrue(a <= v)
        self.assertTrue(a >= v)
        
        v.set_deg(12.4)
        self.assertFalse(a == v)
        self.assertTrue(a != v)
        self.assertTrue(a < v)
        self.assertTrue(a <= v)
        self.assertTrue(v > a)
        self.assertTrue(v >= a)

    def test_compound_operators(self):
        a = angle.new_deg(12.2)
        v = angle.new_deg(12.3)
        
        c = a + v
        self.assertAlmostEqual(c.deg(), 24.5, places=4)
        
        c = a - v
        self.assertAlmostEqual(c.deg(), -0.1, places=4)
        
        self.assertAlmostEqual((a / 2).deg(), 6.1, places=4)
        self.assertTrue(angle.new_deg(6.1) == (a / 2))

    def test_conversion(self):
        a = 0.0
        while a < 360:
            b = angle.new_deg(a)
            a_rad = (a * math.pi) / 180
            
            AssertTolerance.assertAlmostEqual(self, a_rad, b.rad())
            AssertTolerance.assertAlmostEqual(self, a, b.deg())
            a += 0.0001

if __name__ == '__main__':
    unittest.main()

File: libdriveless/tests/src/test_angle.cpp
#include <gtest/gtest.h>
#include "../../include/angle.h"
#include "test_utils.h"
#include <cmath>

TEST(Angle, TestBooleanOperators)
{
    angle a = angle::deg(12.2);
    angle v = angle::deg(12.2);
    
    ASSERT_TRUE(a == v);
    ASSERT_TRUE(a <= v);
    ASSERT_TRUE(a >= v);
    v.setDeg(12.4);
    ASSERT_FALSE(a == v);
    ASSERT_TRUE(a != v);
    ASSERT_TRUE(a < v);
    ASSERT_TRUE(a <= v);
    ASSERT_TRUE(v > a);
    ASSERT_TRUE(v >= a);
}

TEST(Angle, TestCompoundOperators)
{
    angle a = angle::deg(12.2);
    angle v = angle::deg(12.3);
    
    angle c = a + v;
    ASSERT_DEQ(24.5, c.deg());

    c = a - v;
    ASSERT_DEQ(-0.1, c.deg());

    ASSERT_DEQ(6.1, (a / 2).deg());
    ASSERT_TRUE(angle::deg(6.1) == (a / 2));
}


TEST(Angle, TestConversion)
{
    double a = 0, a_rad = 0;

    while (a < 360) {
        auto b = angle::deg(a);
        a_rad = (a * PI)/180;
        ASSERT_DEQ(a_rad, b.rad());
        ASSERT_DEQ(a, b.deg());
        a += 0.0001;
    }    
}

TEST(Angle, TestRadianConversion)
{
    ASSERT_FLOAT_EQ(angle::deg(0).rad(), 0);
    ASSERT_FLOAT_EQ(angle::deg((double)45/2).rad(), PI/8);
    ASSERT_FLOAT_EQ(angle::deg(45).rad(), PI/4);
    ASSERT_FLOAT_EQ(angle::deg(67.5).rad(), 3*PI/8);
    ASSERT_FLOAT_EQ(angle::deg(90).rad(), PI/2);
}